torba/baseaccount.py

```diff
@@ -108,13 +108,22 @@ def get_private_key(self, index: int) -> PrivateKey:
 
     @defer.inlineCallbacks
     def generate_keys(self, start: int, end: int) -> defer.Deferred:
-        new_keys = []
+        keys_batch, final_keys = [], []
         for index in range(start, end+1):
-            new_keys.append((index, self.public_key.child(index)))
-        yield self.db.add_keys(
-            self.account, self.chain_number, new_keys
-        )
-        return [key[1].address for key in new_keys]
+            keys_batch.append((index, self.public_key.child(index)))
+            if len(keys_batch) % 180 == 0:
+                yield self.db.add_keys(
+                    self.account, self.chain_number, keys_batch
+                )
+                final_keys.extend(keys_batch)
+                keys_batch.clear()
+        if keys_batch:
+            yield self.db.add_keys(
+                self.account, self.chain_number, keys_batch
+            )
+            final_keys.extend(keys_batch)
+            keys_batch.clear()
+        return [key[1].address for key in final_keys]
```

Comment:
There should be a simpler way to do this solely in a single loop; without a cleanup conditional after the loop.

Maybe this for inspiration, it loops over a header file in chunks:

https://github.com/lbryio/torba/blob/64a626cbb56164eaeb0a0530bad5258445c984b1/torba/baseheader.py#L187-L192

You can do a loop within a loop, the other loop loops over chunks and the inner loop loops over individual addresses.

Reply:
does aggregating it under a single conditional on the loop solves as you meant? (see last commit)